### PR TITLE
Use IndexOfAnyValues in Uri

### PR DESCRIPTION
**Pull request by MihaZupan**
Contributes to #78204

There are more places we can use `IndexOfAnyValues` in `System.Uri`. I'll address those in future PRs to keep them reviewable, this is complex enough as it is.

|        Method | Toolchain |                     Input |     Mean |   Error | Ratio | Allocated |
|-------------- |---------- |-------------------------- |---------:|--------:|------:|----------:|
| NewUriIdnHost |      main | `http://aaaaaaaa(...)` [50] | 428.2 ns | 2.98 ns |  1.00 |     240 B |
| NewUriIdnHost |        pr | `http://aaaaaaaa(...)` [50] | 314.0 ns | 1.31 ns |  0.73 |     240 B |
|               |           |                           |          |         |       |           |
| NewUriIdnHost |      main | `http://www.Micr(...)` [28] | 254.3 ns | 1.29 ns |  1.00 |     248 B |
| NewUriIdnHost |        pr | `http://www.Micr(...)` [28] | 232.2 ns | 0.67 ns |  0.91 |     192 B |
|               |           |                           |          |         |       |           |
| NewUriIdnHost |      main | `http://www.micr(...)` [28] | 240.6 ns | 1.09 ns |  1.00 |     192 B |
| NewUriIdnHost |        pr | `http://www.micr(...)` [28] | 224.2 ns | 0.64 ns |  0.93 |     192 B |

<details>
<summary>Benchmark code</summary>

```c#
[MemoryDiagnoser]
public class UriBenchmarks
{
    public static IEnumerable<string> Inputs()
    {
        yield return "http://www.microsoft.com/foo";
        yield return "http://www.Microsoft.com/foo";
        yield return $"http://{string.Concat(Enumerable.Repeat(new string('a', 12) + '.', 3))}/foo";
    }

    [ParamsSource(nameof(Inputs))]
    public string Input;

    [Benchmark]
    public string NewUriIdnHost() => new Uri(Input).IdnHost;
}
```

</details>

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 1]
```diff
@@ -12,216 +13,174 @@ namespace System
...
-        //
-        // Returns:
-        //  bool
-        //
-        //  Remarks: Optimized for speed as a most common case,
-        //           MUST NOT be used unless all input indexes are verified and trusted.
-        //
-
-        internal static unsafe bool IsValid(char* name, int pos, ref int returnedEnd, ref bool notCanonical, bool notImplicitFile)
+        public static bool IsValid(ReadOnlySpan<char> hostname, bool iri, bool notImplicitFile, out int length)
```
Comment (MihaZupan) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:101]: The `iri` parameter is always constant, so we could handle the differences with a generic, e.g.
```c#
private interface IIsValidImpl
{
    static abstract int IndexOfInvalidOrDelimiterChar(ReadOnlySpan<char> chars);
    static abstract bool CheckFirstLabelCharacter(char c);
    static abstract int DotIndex(ReadOnlySpan<char> chars);
    static abstract int CalculateLabelLength(ReadOnlySpan<char> label);
}
```
but it seemed like it was adding more code than it's worth (although arguably more readable?).

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 2]
```diff
@@ -12,216 +13,194 @@ namespace System
...
-            string? res = null;
-
-            for (int i = end - 1; i >= start; --i)
+            // Do a quick search for the colon or uppercase letters
+            int index = str.AsSpan(start, end - start).LastIndexOfAny(s_asciiLetterUpperOrColonChars);
+            if (index >= 0)
             {
-                if (char.IsAsciiLetterUpper(str[i]))
+                Debug.Assert(!str.AsSpan(start, index).Contains(':'),
+                    "A colon should appear at most once, and must never be followed by letters.");
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:65]: ParseCanonicalName is handed already sanitized data?

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 3]
```diff
@@ -12,216 +13,194 @@ namespace System
...
+                : hostname.IndexOfAnyExcept(s_validChars);
+
+            if (invalidCharOrDelimiterIndex >= 0)
             {
-                char ch = *newPos;
-                if (ch > 0x7f) return false;    // not ascii
-                if (ch < 'a') // Optimize for lower-case letters, which make up the majority of most Uris, and which are all greater than symbols checked for below
+                char c = hostname[invalidCharOrDelimiterIndex];
+
+                if (c == '/' || c == '\\' || (notImplicitFile && (c == ':' || c == '?' || c == '#')))
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:111]: Nit: as long as you're changing this, you could use newer syntax:
```suggestion
                if (c is '/' or '\\' || (notImplicitFile && (c is ':' or '?' or '#')))
```

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 4]
```diff
@@ -12,216 +13,194 @@ namespace System
...
+                }
+                else
+                {
+                    length = 0;
+                    return false;
                 }
             }
 
-            if (end == curPos)
+            length = hostname.Length;
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:122]: Either move this to after the IsEmpty check or change the IsEmpty check to be length == 0?

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 5]
```diff
@@ -12,216 +13,194 @@ namespace System
...
+            //  Determines whether a string is a valid domain name label. In keeping
+            //  with RFC 1123, section 2.1, the requirement that the first character
+            //  of a label be alphabetic is dropped. Therefore, Domain names are
+            //  formed as:
+            //
+            //      <label> -> <alphanum> [<alphanum> | <hyphen> | <underscore>] * 62
 
-                //check the label start/range
-                if (curPos == newPos || newPos - curPos > 63 || !IsASCIILetterOrDigit(*curPos++, ref notCanonical))
+            // We already verified the content, now verify the lenghts of individual labels
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:136]:
```suggestion
            // We already verified the content, now verify the lengths of individual labels
```

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs [hunk 6]
```diff
@@ -12,216 +13,194 @@ namespace System
...
                 {
-                    end = newPos;
-                    break;
-                }
-            }
+                    ReadOnlySpan<char> label = hostname.Slice(0, labelLength);
+                    if (!IsAscii(label))
+                    {
+                        // s_iriInvalidAsciiChars confirmed everything in [0, 7F] range.
+                        // Chars in [80, A0) range are also invalid, check for them now.
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:157]:
```suggestion
                        // Chars in [80, 9F] range are also invalid, check for them now.
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/DomainNameHelper.cs:65]: Yes

**Pull request merged**